<?xml version="1.0" encoding="UTF-8"?>
<robot name="crackvac">
    <!-- materials -->
    <material name="grey">
       <color rgba="0.7 0.7 0.7 1.0"/> 
    </material>
    <material name="blue">
       <color rgba="0 0 0.9 1.0"/> 
    </material>
    <material name="black">
       <color rgba="0 0 0 1.0"/> 
    </material>

    <link name="base_link">
        <visual>
           <geometry>
              <box size="0.4 0.4 0.2"/> 
           </geometry> 
           <origin xyz="0.0 0.0 0.1" rpy="0.0 0.0 0.0"/>
           <material name="grey"/>
        </visual>
        <collision>
            <geometry>
                <box size="0.4 0.4 0.2"/> <!-- length unit: m-->
            </geometry>
        </collision>
        <inertial>
            <mass value="7.0"/> <!-- mass unit: kg-->
            <inertia ixx="0.187" ixy="0.0" ixz="0.0" iyy="0.117" iyz="0.0" izz="0.117"/>
        </inertial>
    </link>    

    <!-- module links -->
    <link name="fl_module">
        <visual>
           <geometry>
                <cylinder radius="0.05" length="0.01"/> 
           </geometry> 
           <origin xyz="0 0 0" rpy="0 0 0"/>
           <material name="blue"/>
        </visual>
    </link>
    <link name="fr_module">
        <visual>
           <geometry>
                <cylinder radius="0.05" length="0.01"/> 
           </geometry> 
           <origin xyz="0 0 0" rpy="0 0 0"/>
           <material name="blue"/>
        </visual>
    </link>
    <link name="bl_module">
        <visual>
           <geometry>
                <cylinder radius="0.05" length="0.01"/> 
           </geometry> 
           <origin xyz="0 0 0" rpy="0 0 0"/>
           <material name="blue"/>
        </visual>
    </link>
    <link name="br_module">
        <visual>
           <geometry>
                <cylinder radius="0.05" length="0.01"/> 
           </geometry> 
           <origin xyz="0 0 0" rpy="0 0 0"/>
           <material name="blue"/>
        </visual>
    </link>

    <!-- wheel links -->
    <link name="fl_wheel">
        <visual>
            <geometry>
                <cylinder radius="0.04" length="0.03"/>
            </geometry>
            <origin xyz="0 0 0" rpy="1.57 0 0"/>
            <material name="black"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.04" length="0.03"/>
            </geometry>
            <origin xyz="0 0 0" rpy="1.57 0 0"/>
        </collision>
        <inertial>
            <mass value="0.079"/>
            <inertia ixx="0.002291" ixy="0.0" ixz="0.0" iyy="0.002291" iyz="0.0" izz="0.0000632"/>
        </inertial>
    </link>
    <link name="fr_wheel">
        <visual>
            <geometry>
                <cylinder radius="0.04" length="0.03"/>
            </geometry>
            <origin xyz="0 0 0" rpy="1.57 0 0"/>
            <material name="black"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.04" length="0.03"/>
            </geometry>
            <origin xyz="0 0 0" rpy="1.57 0 0"/>
        </collision>
        <inertial>
            <mass value="0.079"/>
            <inertia ixx="0.002291" ixy="0.0" ixz="0.0" iyy="0.002291" iyz="0.0" izz="0.0000632"/>
        </inertial>
   </link>
    <link name="bl_wheel">
        <visual>
            <geometry>
                <cylinder radius="0.04" length="0.03"/>
            </geometry>
            <origin xyz="0 0 0" rpy="1.57 0 0"/>
            <material name="black"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.04" length="0.03"/>
            </geometry>
            <origin xyz="0 0 0" rpy="1.57 0 0"/>
        </collision>
        <inertial>
            <mass value="0.079"/>
            <inertia ixx="0.002291" ixy="0.0" ixz="0.0" iyy="0.002291" iyz="0.0" izz="0.0000632"/>
        </inertial>
    </link>
    <link name="br_wheel">
        <visual>
            <geometry>
                <cylinder radius="0.04" length="0.03"/>
            </geometry>
            <origin xyz="0 0 0" rpy="1.57 0 0"/>
            <material name="black"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.04" length="0.03"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="0.079"/>
            <inertia ixx="0.002291" ixy="0.0" ixz="0.0" iyy="0.002291" iyz="0.0" izz="0.0000632"/>
        </inertial>
    </link>

    <!-- wheel & steer module joints -->
    <joint name="fl_steer_joint" type="continuous">
        <parent link="base_link"/>
        <child link="fl_module"/>
        <origin xyz="0.15 -0.15 0" rpy="0 0 0"/> 
        <axis xyz="0 0 1"/>
    </joint>
    <joint name="fl_wheel_joint" type="continuous">
       <parent link="fl_module"/>
       <child link="fl_wheel"/>
       <origin xyz="0 0 0" rpy="0.0 0.0 0.0"/>
       <axis xyz="0 1 0"/>
    </joint>

    <joint name="fr_steer_joint" type="continuous">
        <parent link="base_link"/>
        <child link="fr_module"/>
        <origin xyz="0.15 0.15 0" rpy="0 0 0"/> 
        <axis xyz="0 0 1"/>
    </joint>
    <joint name="fr_wheel_joint" type="continuous">
       <parent link="fr_module"/>
       <child link="fr_wheel"/>
       <origin xyz="0 0 0" rpy="0.0 0.0 0.0"/>
       <axis xyz="0 1 0"/>
    </joint>

    <joint name="bl_steer_joint" type="continuous">
        <parent link="base_link"/>
        <child link="bl_module"/>
        <origin xyz="-0.15 -0.15 0" rpy="0 0 0"/> 
        <axis xyz="0 0 1"/>
    </joint>
    <joint name="bl_wheel_joint" type="continuous">
       <parent link="bl_module"/>
       <child link="bl_wheel"/>
       <origin xyz="0 0 0" rpy="0.0 0.0 0.0"/>
       <axis xyz="0 1 0"/>
    </joint>

    <joint name="br_steer_joint" type="continuous">
        <parent link="base_link"/>
        <child link="br_module"/>
        <origin xyz="-0.15 0.15 0" rpy="0 0 0"/> 
        <axis xyz="0 0 1"/>
    </joint>
    <joint name="br_wheel_joint" type="continuous">
       <parent link="br_module"/>
       <child link="br_wheel"/>
       <origin xyz="0 0 0" rpy="0.0 0.0 0.0"/>
       <axis xyz="0 1 0"/>
    </joint>
</robot>